expr/eval.go

```diff
@@ -715,179 +712,26 @@ func NewCast(expr Evaluator, typ string) (Evaluator, error) {
-	zv, err := i.expr.Eval(rec)
-	if err != nil {
-		return zng.Value{}, err
-	}
-	v, ok := coerce.ToInt(zv)
-	// XXX better error message
-	if !ok || (i.min != 0 && (v < i.min || v > i.max)) {
-		return zng.Value{}, ErrBadCast
-	}
-	// XXX GC
-	return zng.Value{i.typ, zng.EncodeInt(v)}, nil
-}
-
-type UintCast struct {
-	expr Evaluator
-	typ  zng.Type
-	max  uint64
-}
-
-func (u *UintCast) Eval(rec *zng.Record) (zng.Value, error) {
-	zv, err := u.expr.Eval(rec)
-	if err != nil {
-		return zng.Value{}, err
-	}
-	v, ok := coerce.ToUint(zv)
-	// XXX better error message
-	if !ok || (u.max != 0 && v > u.max) {
-		return zng.Value{}, ErrBadCast
-	}
-	// XXX GC
-	return zng.Value{u.typ, zng.EncodeUint(v)}, nil
-}
-
-type Float64Cast struct {
-	expr Evaluator
-}
-
-func (i *Float64Cast) Eval(rec *zng.Record) (zng.Value, error) {
-	zv, err := i.expr.Eval(rec)
-	if err != nil {
-		return zng.Value{}, err
-	}
-	f, ok := coerce.ToFloat(zv)
-	if !ok {
-		return zng.Value{}, ErrBadCast
-	}
-	return zng.Value{zng.TypeFloat64, zng.EncodeFloat64(f)}, nil
-}
-
-type IPCast struct {
-	expr Evaluator
-}
-
-func (i *IPCast) Eval(rec *zng.Record) (zng.Value, error) {
-	zv, err := i.expr.Eval(rec)
-	if err != nil {
-		return zng.Value{}, err
-	}
-	if !zv.IsStringy() {
-		return zng.Value{}, ErrBadCast
-	}
-	ip := net.ParseIP(string(zv.Bytes))
-	if ip == nil {
-		return zng.Value{}, ErrBadCast
-	}
-	// XXX GC
-	return zng.Value{zng.TypeIP, zng.EncodeIP(ip)}, nil
-}
-
-type TimeCast struct {
-	expr Evaluator
-}
-
-func (t *TimeCast) Eval(rec *zng.Record) (zng.Value, error) {
-	zv, err := t.expr.Eval(rec)
-	if err != nil {
-		return zng.Value{}, err
-	}
-	if zng.IsFloat(zv.Type.ID()) {
-		f, _ := zng.DecodeFloat64(zv.Bytes)
-		ts := nano.FloatToTs(f)
-		// XXX GC
-		return zng.Value{zng.TypeTime, zng.EncodeTime(ts)}, nil
-	}
-	ns, ok := coerce.ToInt(zv)
-	if !ok {
-		return zng.Value{}, ErrBadCast
-	}
-	return zng.Value{zng.TypeTime, zng.EncodeTime(nano.Ts(ns))}, nil
-}
-
-type StringCast struct {
-	expr Evaluator
-	typ  zng.Type
-}
-
-func (s *StringCast) Eval(rec *zng.Record) (zng.Value, error) {
-	zv, err := s.expr.Eval(rec)
-	if err != nil {
-		return zng.Value{}, err
-	}
-	if zv.Type.ID() == zng.IdBytes {
-		return zng.Value{s.typ, zng.EncodeString(string(zv.Bytes))}, nil
-	}
-	if enum, ok := zv.Type.(*zng.TypeEnum); ok {
-		selector, _ := zng.DecodeUint(zv.Bytes)
-		element, err := enum.Element(int(selector))
-		if err != nil {
-			return zng.NewError(err), nil
-		}
-		return zng.Value{s.typ, zng.EncodeString(element.Name)}, nil
-	}
-	//XXX here, we need to create a human-readable string rep
-	// rather than a tzng encoding, e.g., for time, an iso date instead of
-	// ns int.  For now, this works for numbers and IPs.  We will fix in a
-	// subsequent PR (see issue #1603).
-	result := zv.Type.StringOf(zv.Bytes, zng.OutFormatUnescaped, false)
-	return zng.Value{s.typ, zng.EncodeString(result)}, nil
-}
-
-type BytesCast struct {
-	expr Evaluator
+type EvalCast struct {
```

Comment:
Not sure if you're planning to use this outside package `eval`, but if not, maybe rename to `evalCast`.

Reply:
👍